### ISSUE TITLE
Memory Consumption Optimization for Cuda-based FloodFill Algorithm

### ISSUE THREAD

**Issue by KhoiDOO** [KhoiDOO/conquer3d#5]
# Problems:
- FloodFill algorithm cost x4 amount of memory compared to fast winding number and pseudo normal.

# Solution
- Use a less consumped datatype for the mask.
- Tiled voxels for masking instead of global masking.

**Comment by KhoiDOO** [comment]
### Resolved in [v0.6.3](https://github.com/KhoiDOO/conquer3d/releases/tag/v0.6.3): Unified Coarse-to-Fine Hierarchical Volumetric Flood Fill Engine (`sign_mode=5`)

---

### 1. API Usage & Executable Python Code Snippets

```python
import torch
import conquer3d
from conquer3d.data.assets import Armadillo

# 1. Load 3D Triangle Mesh Asset
mesh = Armadillo(device="cuda")

# 2. Build Coarse-to-Fine Flood Fill Data Structure
# Automatically computes optimal macro-block divisor (8x8x8) and coarse/fine masks
cf_data = mesh.build_flood_fill_cf_data(grid_res=[512, 512, 512], block_size=[8, 8, 8])

# 3. High-Performance Point Query with sign_mode=5
# Evaluates signed distance values with O(1) early-exit on exterior/interior macro-blocks
query_pts = torch.rand((1000000, 3), device="cuda") * 2.0 - 1.0
sdf_values = mesh.query_points(query_pts, sign_mode=5)

# 4. Watertight Isosurface Extraction with tmesh2sparse
sparse_grid = conquer3d.conversion.tmesh2sparse(
    mesh,
    grid_res=512,
    sign_mode=5,
    block_size=[8, 8, 8]
)

# 5. Extract Watertight Meshes with Isosurface Extraction Algorithms
verts_mc, faces_mc = conquer3d.ops.marching_cubes(sparse_grid.coords, sparse_grid.values)
verts_dc, faces_dc = conquer3d.ops.dual_contouring(sparse_grid.coords, sparse_grid.values)
verts_dmc, faces_dmc = conquer3d.ops.dual_marching_cubes(sparse_grid.coords, sparse_grid.values)
```

---

### 2. Advantages & Core Capabilities
- **$314.7\times$ Peak VRAM Reduction**:
  - Resolution $512^3$: **$5.58\text{ MB}$ vs $1,152.00\text{ MB}$ ($206.6\times$ reduction)**.
  - Resolution $1024^3$: **$29.28\text{ MB}$ vs $9,216.00\text{ MB}$ ($314.7\times$ reduction)**, saving **$9.18\text{ GB}$ of VRAM** on GPU.
- **$100.000\%$ Topological & Geometric Parity**:
  - Unified 2-Level Multi-Scale Wavefront Graph propagation eliminates all premature coarse interior classifications and macro-block wall blocking.
  - Guaranteed $0$ open boundary edges and identical vertex/face counts compared to dense flood fill across Marching Cubes, Dual Contouring, Dual Marching Cubes, and Marching Tetrahedra.
- **$O(1)$ Early-Exit Point Queries**:
  - BVH point query evaluates coarse macro-blocks first; points in exterior open water or deep interior solid resolve instantly in $1$ memory access without traversing BVH leaf triangles.

---

### 3. Disadvantages & Known Trade-offs
- **Iterative Wavefront Convergence**: Unified multi-scale propagation runs synchronized host loop iterations across coarse and fine kernels until convergence (`changed_flag == 0`).
- **Shared Memory Limit**: Fine intra-block BFS uses `int8_t s_mask[8][8][8]` ($512$ bytes shared memory per CUDA threadblock), capping the ideal macro-block dimension to $8\times 8\times 8$.

---

### 4. Empirical Benchmarks & Performance Metrics

#### Stanford Armadillo ($49,990$ vertices, $99,976$ faces) on NVIDIA RTX GPU

| Resolution | Grid Voxels | Dense VRAM (`sign_mode=3`) | CF VRAM (`sign_mode=5`) | Memory Reduction | Parity Status |
| :--- | :--- | :--- | :--- | :---: | :---: |
| **$512^3$** | $134,217,728$ | $1,152.00\text{ MB}$ | **$5.58\text{ MB}$** | **$206.6\times$** | **$100.000\%$ Exact** |
| **$1024^3$** | $1,073,741,824$ | $9,216.00\text{ MB}$ | **$29.28\text{ MB}$** | **$314.7\times$** | **$100.000\%$ Exact** |

#### Isosurface Extraction Parity (Zero Missing Vertices / Zero Holes)

| Extraction Method | Resolution $512^3$ Vertices | Resolution $512^3$ Faces | Resolution $1024^3$ Vertices | Resolution $1024^3$ Faces |
| :--- | :--- | :--- | :--- | :--- |
| **Marching Cubes (MC)** | $466,196$ | $932,388$ | $1,872,248$ | $3,744,492$ |
| **Dual Contouring (DC)** | $466,174$ | $932,392$ | $1,872,220$ | $3,744,496$ |
| **Dual Marching Cubes (DMC)** | $466,198$ | $932,392$ | $1,872,250$ | $3,744,496$ |
| **Marching Tetrahedra (MT)** | $1,421,782$ | $2,843,560$ | — | — |

---

### 5. Technical Insights & Architectural Decisions
- **Unified 2-Level Multi-Scale Graph Wavefront**:
  1. `coarse_to_coarse_kernel`: Expands exterior water across empty macro-blocks.
  2. `coarse_to_fine_kernel`: Seeds boundary block faces touching coarse exterior water via BVH line-of-sight tests.
  3. `fine_intra_block_bfs_kernel`: Propagates water in shared memory avoiding SAT-colliding triangle boundaries.
  4. `fine_to_fine_inter_block_kernel`: Exchanges water across adjacent boundary block faces.
  5. `fine_to_coarse_kernel`: Seeds unreached empty coarse blocks from fine water voxels.
- **Exact Separating Axis Theorem (SAT)**: Macro-blocks are filtered with exact 3D triangle-box intersection tests (`Triangle::is_voxel_intersect`), preventing false boundary classification.
- **Line-of-Sight Hardware BVH Ray-Segment Tests**: All inter-scale transitions test segment intersection against the BVH to prevent water leakage into interior solid cavities while ensuring complete exterior flow.
